### PR TITLE
py-scipy: add v1.12 and v1.13

### PR DESCRIPTION
**Pull request by adamjstewart**
https://github.com/scipy/scipy/releases/tag/v1.12.0

Only tests that failed for me:
```
________________ TestZlibInputStream.test_all_data_read_overlap ________________
/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/py-scipy-1.12.0-mtnrdf4mechwsgcttr6j372y7cllxdun/lib/python3.11/site-packages/scipy/io/matlab/tests/test_streams.py:205: in test_all_data_read_overlap
    assert_(compressed_data_len == BLOCK_SIZE + 2)
E   AssertionError
        COMPRESSION_LEVEL = 6
        compressed_data = b'x\x9c\xec\xdd\x05[U\xdb\x1a\x06PB\x04\x04\xe9n\x01i\x94nP\x1aI\x91\x94\x10A\x1a\x14\x14\x01\xf5\xdc\xee\xee\xee\xee\...bb\xff\xdd\xff\xee\x7f\xf7\xbf\xfb\xdf\xfd\xef\xfew\xff\xbb\xff\xdd\xff\xee\x7f\xf7\xbf_\xfe\xcf\xff\xfe\x0fl\xcc\xd1^'
        compressed_data_len = 251004
        data       = b'\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x...da\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7'
        self       = <scipy.io.matlab.tests.test_streams.TestZlibInputStream object at 0x11a3ef810>
_____________ TestZlibInputStream.test_all_data_read_bad_checksum ______________
/Users/Adam/spack/opt/spack/darwin-sonoma-m2/apple-clang-15.0.0/py-scipy-1.12.0-mtnrdf4mechwsgcttr6j372y7cllxdun/lib/python3.11/site-packages/scipy/io/matlab/tests/test_streams.py:221: in test_all_data_read_bad_checksum
    assert_(compressed_data_len == BLOCK_SIZE + 2)
E   AssertionError
        COMPRESSION_LEVEL = 6
        compressed_data = b'x\x9c\xec\xdd\x05[U\xdb\x1a\x06PB\x04\x04\xe9n\x01i\x94nP\x1aI\x91\x94\x10A\x1a\x14\x14\x01\xf5\xdc\xee\xee\xee\xee\...bb\xff\xdd\xff\xee\x7f\xf7\xbf\xfb\xdf\xfd\xef\xfew\xff\xbb\xff\xdd\xff\xee\x7f\xf7\xbf_\xfe\xcf\xff\xfe\x0fl\xcc\xd1^'
        compressed_data_len = 251004
        data       = b'\x00\x01\x02\x03\x04\x05\x06\x07\x08\t\n\x0b\x0c\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\x1b\x1c\x1d\x...da\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7'
        self       = <scipy.io.matlab.tests.test_streams.TestZlibInputStream object at 0x11a48b310>
=========================== short test summary info ============================
FAILED ../io/matlab/tests/test_streams.py::TestZlibInputStream::test_all_data_read_overlap
FAILED ../io/matlab/tests/test_streams.py::TestZlibInputStream::test_all_data_read_bad_checksum
= 2 failed, 58897 passed, 2722 skipped, 237 xfailed, 16 xpassed in 671.22s (0:11:11) =
```
Let me know if I need to add any other deps to avoid the 2722 that were skipped.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]: @rgommers in the past I know we talked about leaving these upper bounds open-ended. However, Spack now has support for separate concretization of build-only dependencies similar to pip. So I don't think it's necessary, we can pin to exact versions now if we want.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]: Not documented in pyproject.toml, but I needed to add a newer version of hypothesis to get things working. For older versions of hypothesis, I got the following error:
```
TypeError: from_dtype() got an unexpected keyword argument 'allow_subnormal'
```
Might be worth documenting in pyproject.toml.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: What exactly is the benefit of an upper bound and why is that better at build time than at runtime?

Note that pip does NOT pin build time dependencies by default and the only way to do it is to release new tagged upstream point releases every time you want to change a pin. So I would argue there is precisely zero prior art here, even if spack contains the technological capability to update pins without new point releases of e.g. a package build ID.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: That's not a question for me, it's a question for the scipy devs. Most build dependencies have historically been pinned to the latest release for "stability" reasons. I'm just trying to match that in Spack.

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: Well, this goes back to the pinning method being part of the project source code and consequently part of the release tag as well.

- It's not possible to optimistically allow newer versions on the assumption that it most likely works, and adapt to the odd case of it not working, by adding a pin. You get one chance to publish your source code tarball to PyPI, and that's at the initial time of release.
- for cibuildwheel, you may want to impose specific restrictions, most famously the use of oldest-supported-numpy / a lower bound on numpy, which aren't even relevant *at all* to people building numpy for personal use, only relevant to cibuildwheel. But the PyPA ecosystem doesn't really have an official pinning infrastructure, and there's no official way to pin dependencies "for cibuildwheel", so lower/upper bounds in build dependencies in pyproject.toml is used as a "pragmatic fallback" to get the official cibuildwheel artifact builds to produce predictable results that support a wide range of numpy ABIs.

Spack doesn't operate under any of these restrictions.

Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: > However, Spack now has support for separate concretization of build-only dependencies similar to pip.

Nice to see this improvement!

> Well, this goes back to the pinning method being part of the project source code and consequently part of the release tag as well.
> [...] Spack doesn't operate under any of these restrictions.

Yes, you're correct about the rationale upstream @eli-schwartz. I really wish PyPI had a better model here, but it doesn't. So for Spack the choice here is:
1. do I simply copy the upstream upper bounds, and loosen them in case there is an issue with an _older_ version of a build dependency
2. do I ignore the upstream upper bounds, and add them whenever there is an issue with a _newer_ version of a build dependency

Given that build deps are concretized separately, the choice only matters for building `py-scipy`. And I don't think the answer matters all that much, it's just whatever is expected to be less effort. And that's hard to tell ahead of time; it's probably similar for both cases. So I'm fine with either choice.

Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: > there is precisely zero prior art here

This isn't entirely true by the way. I stumbled over https://doc.rust-lang.org/cargo/faq.html#can-libraries-use--as-a-version-for-their-dependencies recently - interesting to see that Cargo by design does not allow not having any bounds. And if you browse dependency info on crates.io, `^` is used pretty consistently.

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: Heh. To be fair by "here" I meant "derived from the pip comparison".

(Also I disagree with cargo's reasoning. There are some notable projects that offer an eternal compatibility guarantee: all new versions implement the API of all old versions, updates shall only ever add APIs and not remove them. e.g. glibc has done this for several decades now.)

Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: > To be fair by "here" I meant "derived from the pip comparison".

Okay, then fair enough.

> Also I disagree with cargo's reasoning

Yeah, I think it's not that hard to produce a counter-example. That said, they're correct for >=99% of the time. `glibc` is about as stable as it gets - but still, it's got lots of bugs, and NumPy contains a bunch of workarounds for them. So definitely not every version works; one needs at least a lower bound for nontrivial usage. Even bug fixes in newer versions may be bc-breaking.

Comment (eli-schwartz) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: (... although for bugs you'd rather mark a specific version as `!` rather than using pinned bounds)

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: > Spack doesn't operate under any of these restrictions.

This is a very valid point.

> So for Spack the choice here is:

And a great summary of the pros and cons of each.

In that sense, I wonder if the lower bound is even a real lower bound. Should we leave off all version requirements until we know for certain that they are real?

I'm fine with both options honestly. @rgommers has been great about responding to Spack issues for numpy/scipy, so I'm not particularly worried about not keeping up-to-speed with issue reports. I'll let y'all decide which you prefer.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: https://github.com/scipy/scipy/pull/19942

**Comment by adamjstewart** [comment]
More importantly, we're seeing an issue in CI when building scipy:
```
  Program pythran found: NO
  ../meson.build:140:12: ERROR: Program 'pythran' not found or not executable
```
Any idea why meson wouldn't be able to find pythran?

**Comment by rgommers** [comment]
> Any idea why meson wouldn't be able to find pythran?

Pythran 0.15.0 switched to `pyproject.toml`. If the `pythran` CLI tool is going missing like this, I bet it's an issue with how the entrypoint is handled. It's quite standard here: https://github.com/serge-sans-paille/pythran/blob/bdae6a7a5204d11988da0c04a9f957c1f4cfe356/pyproject.toml#L32-L34

Pythran doesn't seem to be packaged in Spack anymore? Latest version at https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/py-pythran/package.py is `0.12.2`, and it's failing on `0.15.0`.

**Comment by adamjstewart** [comment]
I updated it to 0.15.0 in this PR.

**Comment by rgommers** [comment]
> I updated it to 0.15.0 in this PR.

Ah of course. I don't see anything unusual there, and it works for you locally. So not quite sure. It must have something to do with entrypoint handling I think.

> Let me know if I need to add any other deps to avoid the 2722 that were skipped.

It's probably mostly the `scipy.special` tests that compare against `mpmath`. Those are pretty slow, but if you want to add `mpmath` as a test dependency that should reduce the skips. Probably needs `gmpy2` as well then to not be extra slow.

There's a few other super slow tests that aren't run by default, and a few that need `matplotlib`. But those really aren't worth bothering with probably.

The `TestZlibInputStream` are fairly harmless but probably worth reporting upstream. Likely depends on the `zlib` version in Spack.

**Comment by adamjstewart** [comment]
Any ideas for debugging the pythran issue?

**Comment by rgommers** [comment]
> Any ideas for debugging the pythran issue?

Is there a way to debug on CI? I checked the Pythran issue tracker, but nothing was reported there from another distro or by an end user. The only other thing I can think of is double/triple checking `pip` and `setuptools` versions, to ensure it's not a bug in an older version of one of those packages.

**Comment by adamjstewart** [comment]
The failing job output contains the following instructions to reproduce:
```
To reproduce this build locally, run:
    spack ci reproduce-build https://gitlab.spack.io/api/v4/projects/2/jobs/9836998/artifacts [--working-dir <dir>] [--autostart]
If this project does not have public pipelines, you will need to first:
    export GITLAB_PRIVATE_TOKEN=<generated_token>
... then follow the printed instructions.
```
This uses our docker images, so it should be possible to reproduce the issue exactly. Not sure if this will work on macOS arm64 though, can anyone try on Linux x86_64?

**Comment by boeschf** [comment]
FYI, I get this compiler error on an arm64 system (aarch64, neoverse v2):
```
[566/1610] Compiling Fortran object scipy/special/libspecfun.a.p/specfun_specfun.f.o
FAILED: scipy/special/libspecfun.a.p/specfun_specfun.f.o
gfortran -Iscipy/special/libspecfun.a.p -Iscipy/special -I../scipy/special -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -std=legacy -O3 -Wno-conversion -fPIC -Wno-maybe-uninitialized -Jscipy/special/libspecfun.a.p -o scipy/special/libspecfun.a.p/specfun_specfun.f.o -c ../scipy/special/specfun/specfun.f
during GIMPLE pass: vect
../scipy/special/specfun/specfun.f:6257:24:

 6257 |         SUBROUTINE CJYNB(N,Z,NM,CBJ,CDJ,CBY,CDY)
      |                        ^
internal compiler error: in compute_live_loop_exits, at tree-ssa-loop-manip.cc:247
 ```
 Probably unrelated to the issues above, though.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: > In that sense, I wonder if the lower bound is even a real lower bound. Should we leave off all version requirements until we know for certain that they are real?

Almost all lower bounds are actual bounds. Usually what happens is that a lower bound is bumped when a real-world bug shows that the older version doesn't work. The only thing you cannot tell from the metadata is whether that bug is a general incompatibility (e.g., the package started using a new feature in the dependency that isn't supported in versions below the lower bound) or it's a platform-specific bug.

> I'll let y'all decide which you prefer.

Now that Spack separately concretizes build deps, really either one works. So I don't have a super strong preference either way. However, I do know that the majority of distro maintainers prefer no upper bounds (as does @eli-schwartz in this discussion), so I suggest we go with that. To be explicit, I propose the following policy for taking over version constraints from metadata in `pyproject.toml` files:

For `build-system.requires`:

1. Do take over lower bounds
2. Do not take over upper bounds, unless there is a specific comment in pyproject.toml that this upper bound is necessary for reason X, and that reason applies in Spack

For `project.dependencies`: take over all constraints exactly.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: Works for me. Let me update both PRs with this new policy.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py:87]: I'm guessing these aren't real upper bounds?

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-scipy/package.py]: Not sure which of these upper bounds are strict, but I do remember some of these being required.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py:87]: Indeed, I don't think they are.

File: var/spack/repos/builtin/packages/py-scipy/package.py
Comment (rgommers) [reply var/spack/repos/builtin/packages/py-scipy/package.py]: Yep, these I would not touch. `setuptools` 60 was definitely breaking everything, not sure about the exact upper bounds for the rest.

**Review by rgommers — APPROVED** [review]
These changes LGTM, thanks Adam.

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Comment by adamjstewart** [comment]
Back to the same error we were seeing before:
```
  ../meson.build:140:12: ERROR: Program 'pythran' not found or not executable
```

**Comment by boeschf** [comment]
> FYI, I get this compiler error on an arm64 system (aarch64, neoverse v2):
> 
> ```
> [566/1610] Compiling Fortran object scipy/special/libspecfun.a.p/specfun_specfun.f.o
> FAILED: scipy/special/libspecfun.a.p/specfun_specfun.f.o
> gfortran -Iscipy/special/libspecfun.a.p -Iscipy/special -I../scipy/special -fdiagnostics-color=always -DNDEBUG -D_FILE_OFFSET_BITS=64 -Wall -std=legacy -O3 -Wno-conversion -fPIC -Wno-maybe-uninitialized -Jscipy/special/libspecfun.a.p -o scipy/special/libspecfun.a.p/specfun_specfun.f.o -c ../scipy/special/specfun/specfun.f
> during GIMPLE pass: vect
> ../scipy/special/specfun/specfun.f:6257:24:
> 
>  6257 |         SUBROUTINE CJYNB(N,Z,NM,CBJ,CDJ,CBY,CDY)
>       |                        ^
> internal compiler error: in compute_live_loop_exits, at tree-ssa-loop-manip.cc:247
> ```
> 
> Probably unrelated to the issues above, though.

just to let you know, this^^ was caused by a [compiler bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111478) and has nothing to do with the issue at hand here.

**Comment by adamjstewart** [comment]
@rgommers any ideas what to do to get this to build?

**Comment by adamjstewart** [comment]
@rgommers looks like 1.13 has the same issue as 1.12 and won't build in CI.

**Comment by rgommers** [comment]
I can now reproduce it in a docker container with `spack ci`, but the `meson-log.txt` with the details is gone. Once I'm in the docker container and go to `/tmp/root/spack-stage/spack-stage-py-scipy-1.13.0-5kad4frf...`, it not quite clear to me how to trigger the rebuild. Maybe I'm being dense, but how do I activate the build env and trigger a rebuild?

**Comment by adamjstewart** [comment]
I'm honestly not sure either. I thought we (I?) wrote a PR to make sure all `archive_files` make it into `artifacts.zip`, but maybe that was only to make sure they end up in `<prefix>/.spack`. @scottwittenburg @kwryankrattiger can you help advise here?

**Comment by rgommers** [comment]
Okay it is there, in `spack-src`:
```docker
$ docker exec -it <container-id> bash
# cd /tmp/root/spack-stage/spack-stage-py-scipy-1.13.0-5kad4frf7pg3pl5elfumwghbs5u4gamf
# vim spack-src/build/meson-logs/meson-log.tx
```
but there's nothing informative there:
```
Program pythran found: NO

../meson.build:138:12: ERROR: Program 'pythran' not found or not executable
```

That line 138 is:
```meson
pythran = find_program('pythran', native: true, version: '>=0.14.0')
```

The `pythran` executable should be installed into `<build-env-prefix>/bin/`. Looking at the [py-pythran build log](https://gitlab.spack.io/spack/spack/-/jobs/10600798) it shows:
```
Successfully built UNKNOWN
Installing collected packages: UNKNOWN
Successfully installed UNKNOWN-0.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Removed build tracker: '/tmp/pip-build-tracker-8fgft3df'
==> py-pythran: Successfully installed py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo
```

<details>
<summary>Full output</summary>

```
==> Installing py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo [1/1]
==> Fetching https://files.pythonhosted.org/packages/source/p/pythran/pythran-0.15.0.tar.gz
==> Ran patch() for py-pythran
==> py-pythran: Executing phase: 'install'
==> [2024-04-03-10:22:02.172567] '/home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/python-3.11.7-ble6bsz2e6fjpfuhj2hobe7wg5aoi66b/bin/python3.11' '-m' 'pip' '-vvv' '--no-input' '--no-cache-dir' '--disable-pip-version-check' 'install' '--no-deps' '--ignore-installed' '--no-build-isolation' '--no-warn-script-location' '--no-index' '--prefix=/home/software/spack/[padded-to-256-chars]/morepadding/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo' '.'
Using pip 23.1.2 from /home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/py-pip-23.1.2-f4ft2ge2r7ze5htwrpioudmremwnviwy/lib/python3.11/site-packages/pip (python 3.11)
Non-user install due to --prefix or --target option
Ignoring indexes: https://pypi.org/simple
Created temporary directory: /tmp/pip-build-tracker-8fgft3df
Initialized build tracking at /tmp/pip-build-tracker-8fgft3df
Created build tracker: /tmp/pip-build-tracker-8fgft3df
Entered build tracker: /tmp/pip-build-tracker-8fgft3df
Created temporary directory: /tmp/pip-install-qf9ay7uq
Created temporary directory: /tmp/pip-ephem-wheel-cache-s7igyb6t
Processing /tmp/root/spack-stage/spack-stage-py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo/spack-src
  Added file:///tmp/root/spack-stage/spack-stage-py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo/spack-src to build tracker '/tmp/pip-build-tracker-8fgft3df'
  Created temporary directory: /tmp/pip-modern-metadata-68gmpe4f
  Preparing metadata (pyproject.toml): started
  Running command Preparing metadata (pyproject.toml)
  running dist_info
  creating /tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info
  writing /tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/PKG-INFO
  writing dependency_links to /tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/dependency_links.txt
  writing top-level names to /tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/top_level.txt
  writing manifest file '/tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/SOURCES.txt'
  reading manifest file '/tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  adding license file 'AUTHORS'
  writing manifest file '/tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.egg-info/SOURCES.txt'
  creating '/tmp/pip-modern-metadata-68gmpe4f/UNKNOWN.dist-info'
  Preparing metadata (pyproject.toml): finished with status 'done'
  Source in /tmp/root/spack-stage/spack-stage-py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo/spack-src has version 0.0.0, which satisfies requirement UNKNOWN==0.0.0 from file:///tmp/root/spack-stage/spack-stage-py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo/spack-src
  Removed UNKNOWN==0.0.0 from file:///tmp/root/spack-stage/spack-stage-py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo/spack-src from build tracker '/tmp/pip-build-tracker-8fgft3df'
Created temporary directory: /tmp/pip-unpack-7yxt3qxj
Building wheels for collected packages: UNKNOWN
  Created temporary directory: /tmp/pip-wheel-auqiiqw6
  Destination directory: /tmp/pip-wheel-auqiiqw6
  Building wheel for UNKNOWN (pyproject.toml): started
  Running command Building wheel for UNKNOWN (pyproject.toml)
  running bdist_wheel
  running build
  installing to build/bdist.linux-x86_64/wheel
  running install
  running install_egg_info
  running egg_info
  creating UNKNOWN.egg-info
  writing UNKNOWN.egg-info/PKG-INFO
  writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
  writing top-level names to UNKNOWN.egg-info/top_level.txt
  writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
  reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE'
  adding license file 'AUTHORS'
  writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
  Copying UNKNOWN.egg-info to build/bdist.linux-x86_64/wheel/UNKNOWN-0.0.0-py3.11.egg-info
  running install_scripts
  creating build/bdist.linux-x86_64/wheel/UNKNOWN-0.0.0.dist-info/WHEEL
  creating '/tmp/pip-wheel-auqiiqw6/tmpg749zmkn/UNKNOWN-0.0.0-py3-none-any.whl' and adding 'build/bdist.linux-x86_64/wheel' to it
  adding 'UNKNOWN-0.0.0.dist-info/AUTHORS'
  adding 'UNKNOWN-0.0.0.dist-info/LICENSE'
  adding 'UNKNOWN-0.0.0.dist-info/METADATA'
  adding 'UNKNOWN-0.0.0.dist-info/WHEEL'
  adding 'UNKNOWN-0.0.0.dist-info/top_level.txt'
  adding 'UNKNOWN-0.0.0.dist-info/RECORD'
  removing build/bdist.linux-x86_64/wheel
  Building wheel for UNKNOWN (pyproject.toml): finished with status 'done'
  Created wheel for UNKNOWN: filename=UNKNOWN-0.0.0-py3-none-any.whl size=2311 sha256=5ac6edf7b9db95b86ff9073be1379cddf8044004b8c2de3144c9dd395ef2e0e0
  Stored in directory: /tmp/pip-ephem-wheel-cache-s7igyb6t/wheels/5b/54/00/7456fe498a76cfa205543e8bf3fac474f9b076b54d53e487d8
Successfully built UNKNOWN
Installing collected packages: UNKNOWN
Successfully installed UNKNOWN-0.0.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
Removed build tracker: '/tmp/pip-build-tracker-8fgft3df'
==> py-pythran: Successfully installed py-pythran-0.15.0-qop6eyiyzs6ckc2zllxmcwiihltr6uuo
```

</details>

That can't be right. There's also a generated build info page showing the same: https://cdash.spack.io/build/5377960. No idea why that happens, but fixing that will make `pythran` re-appear I bet.

**Comment by kwryankrattiger** [comment]
> I'm honestly not sure either. I thought we (I?) wrote a PR to make sure all `archive_files` make it into `artifacts.zip`, but maybe that was only to make sure they end up in `<prefix>/.spack`. @scottwittenburg @kwryankrattiger can you help advise here?

Just looking at the code I think we should be copying them, but I think in practice this has not been the case as reported here and by others looking for these logs.

* `ci.py`:
```python
    for file in [job_pkg.log_path, job_pkg.env_mods_path, *job_pkg.builder.archive_files]:
        copy_files_to_artifacts(file, job_log_dir)
```

* `build_systems/meson.py`
```python
    def archive_files(self):
        """Files to archive for packages based on Meson"""
        return [os.path.join(self.build_directory, "meson-logs", "meson-log.txt")]
```

As far as debugging, you should be able to regenerate these logs in the container and open them locally that way. I will do some more investigation as to what is going wrong here.

**Comment by kwryankrattiger** [comment]
> Okay it is there, in spack-src:

Oh great! That is an easy fix!

**Comment by adamjstewart** [comment]
`UNKNOWN==0.0.0` looks very weird indeed. Any idea how they detect name/version for that package? If it requires access to git, we can get that.

**Comment by rgommers** [comment]
I can't reproduce it locally from git or from sdist. I'll note that there are three separate jobs that build `py-pythran`, and two of them look fine. The output for the third one is weird and has:
```
py-setuptools: Successfully installed py-setuptools-59.4.0-obifypzjwsrsy4rjal3coq4e4oty56yf
```
that's probably too old; it may be using some legacy thing - there were lots of changes to `distutils` and build isolation behavior around `setuptools` 60.0-62.0. Updating to a new version will fix it I think.

**Comment by adamjstewart** [comment]
Might hit concretization errors, but we'll at least be out of pythran hell now!

**Comment by adamjstewart** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]

**Comment by adamjstewart** [comment]
Thanks again for the help @rgommers! Next fun experiments will be numpy 2.0 and matplotlib 3.9 (meson backend). Looking forward to seeing how many packages have not yet declared numpy 2 compatibility!

**Comment by rgommers** [comment]
Thank you for all the hard work!

NumPy 2.0 will be a while away I suspect - there's a ton of packages that still need to do releases, and I suspect there will still be laggards once the final 2.0.0 release comes out.

**Pull request merged**
